File: Hex/Hex.java
package Hex;

/**
 * <pre>Coordinate system: (2i, 2j, 2k)
 *    I
 *   / * (5, 4, -1)
 *  /     * (5, 7, 2)
 * o - - J
 *  \ * (0, 3, 3)
 *   \
 *    K
 * </pre>
 * <p>Coordinates: dimension along the line I, J, or K.</p>
 * <p>Line distances: distances to I or K, which are perpendicular to the coordinates.</p>
 */

public class Hex{
    private final double sinOf60 = Math.sqrt(3) / 4;
    private int x;
    private int y;

    // Basic constructors
    /**
     * Default constructor initializing the hex coordinate at (0,0).
     */
    public Hex(){
        // Basic constructor
        this.x = 0;
        this.y = 0;
    }
    /**
     * Constructs a hex at the specified (i, k) coordinates.
     *
     * @param i The i-coordinate.
     * @param k The k-coordinate.
     */
    public Hex(int i, int k){
        // Coordinate constructor
        this.x = i;
        this.y = k;
    }

    // Line constructors (static)
    // Please use those for the game instead of the old constructors
    /**
     * Creates a default hex coordinate at (0,0).
     *
     * @return A new hex instance at the origin.
     */
    public static Hex hex(){
        return new Hex();
    }
    /**
     * Creates a hex coordinate using hexagonal line indices instead of direct coordinates.
     * The hex is shifted accordingly in the coordinate system.
     *
     * @param i The I-line index in the hexagonal coordinate system.
     * @param k The K-line index in the hexagonal coordinate system.
     * @return A new hex coordinate positioned according to the given line indices.
     */
    public static Hex hex(int i, int k){
        return new Hex().shiftI(k).shiftK(i);
    }

    // Raw coordinates
    /**
     * Raw I-coordinate
     * @return The I-coordinate.
     */
    public int I(){
        return x;
    }
    /**
     * Raw J-coordinate
     * @return The J-coordinate.
     */
    public int J(){
        return x + y;
    }
    /**
     * Raw K-coordinate
     * @return The K-coordinate.
     */
    public int K(){
        return y;
    }

    // Lines
    /**
     * Computes the line index along the I-axis in the hexagonal coordinate system.
     * The I-axis represents a diagonal axis in the hexagonal coordinate system.
     *
     * @return The computed I-line index.
     * @see #getLines()
     */
    public int getLineI(){
        return (2*y+x)/3;
    }
    /**
     * Computes the line index along the J-axis in the hexagonal coordinate system.
     * The J-axis represents a diagonal axis in the hexagonal coordinate system.
     *
     * @return The computed J-line index.
     * @see #getLines()
     */
    public int getLineJ(){
        return (x-y)/3;
    }
    /**
     * Computes the line index along the K-axis in the hexagonal coordinate system.
     * The K-axis represents a diagonal axis in the hexagonal coordinate system.
     *
     * @return The computed K-line index.
     * @see #getLines()
     */
    public int getLineK(){
        return (2*x+y)/3;
    }
    /**
     * String representation of the line indices of the hex along all axes
     * <p>Format: {@code {I = i, J = j, K = k}}</p>
     *
     * @return A formatted string representing the hex coordinate indices along I, J, and K axes.
     * @see #getLineI()
     * @see #getLineJ()
     * @see #getLineK()
     */
    public String getLines(){
        return "{I = " + getLineI() + ", J = " + getLineJ() + ", K = " + getLineK() + "}";
    }

    // Line booleans
    /**
     * Determines whether this hex is in the given I-line.
     * The I-line represents a diagonal axis in the hexagonal coordinate system.
     *
     * @param line The I-line value to check.
     * @return True if the hex is in the specified I-line, otherwise false.
     * @see #inLineI(Hex)
     */
    public boolean inLineI(int line){
        return getLineI() == line;
    }
    /**
     * Determines whether this hex is in the given J-line.
     * The J-line represents a diagonal axis in the hexagonal coordinate system.
     *
     * @param line The J-line value to check.
     * @return True if the hex is in the specified J-line, otherwise false.
     * @see #inLineJ(Hex)
     */
    public boolean inLineJ(int line){
        return getLineJ() == line;
    }
    /**
     * Determines whether this hex is in the given K-line.
     * The K-line represents a diagonal axis in the hexagonal coordinate system.
     *
     * @param line The K-line value to check.
     * @return True if the hex is in the specified K-line, otherwise false.
     * @see #inLineK(Hex)
     */
    public boolean inLineK(int line){
        return getLineK() == line;
    }
    /**
     * Determines whether this hex coordinate is in the same I-line as the other hex coordinate.
     * The I-line represents a diagonal axis in the hexagonal coordinate system.
     *
     * @param other The other hex coordinate to compare with.
     * @return True if this hex coordinate is in the same I-line as the other hex coordinate.
     * @see #inLineI(int)
     */
    public boolean inLineI(Hex other){
        return this.getLineI() == other.getLineI();
    }
    /**
     * Determines whether this hex coordinate is in the same J-line as the other hex coordinate.
     * The J-line represents a diagonal axis in the hexagonal coordinate system.
     *
     * @param other The other hex coordinate to compare with.
     * @return True if this hex coordinate is in the same K-line as the other hex coordinate.
     * @see #inLineJ(int)
     */
    public boolean inLineJ(Hex other){
        return this.getLineJ() == other.getLineJ();
    }
    /**
     * Determines whether this hex coordinate is in the same K-line as the other hex coordinate.
     * The K-line represents a diagonal axis in the hexagonal coordinate system.
     *
     * @param other The other hex coordinate to compare with.
     * @return True if this hex coordinate is in the same K-line as the other hex coordinate.
     * @see #inLineK(int)
     */
    public boolean inLineK(Hex other){
        return this.getLineK() == other.getLineK();
    }
    /**
     * Determines whether this hex coordinate is adjacent to another hex.
     * Two hex coordinates are considered adjacent if they share an edge in the hexagonal grid.
     * Use front and back for more precise finding.
     *
     * @param other The other hex to compare with.
     * @return True if the hex coordinates are adjacent, otherwise false.
     * @see #front(Hex)
     * @see #back(Hex)
     */
    public boolean adjacent(Hex other){
        return front(other) || back(other);
    }
    /**
     * Determines if this hex coordinate is in front of another hex coordinate.
     * A hex coordinate is considered "in the front" if it is positioned one step higher in any of the three coordinate axes.
     *
     * @param other The other hex to compare with.
     * @return True if this hex coordinate is one unit higher in I, J, or K, otherwise false.
     * @see #adjacent(Hex)
     * @see #frontI(Hex)
     * @see #frontJ(Hex)
     * @see #frontK(Hex)
     */
    public boolean front(Hex other){
        // adjacent, this is one higher in I, J, or K
        return frontI(other) || frontJ(other) || frontK(other);
    }
    /**
     * Determines if this hex coordinate is behind another hex coordinate.
     * A hex coordinate is considered "behind" if it is positioned one step lower in any of the three coordinate axes.
     *
     * @param other The other hex to compare with.
     * @return True if this hex coordinate is one unit lower in I, J, or K, otherwise false.
     * @see #adjacent(Hex)
     * @see #backI(Hex)
     * @see #backJ(Hex)
     * @see #backK(Hex)
     */
    public boolean back(Hex other){
        // adjacent, this is one lower in I, J, or K
        return backI(other) || backJ(other) || backK(other);
    }
    /**
     * Determines if this hex coordinate is in front of another hex coordinate on the I-axis.
     * A hex coordinate is considered "in the front" if it is positioned one step higher.
     *
     * @param other The other hex to compare with.
     * @return True if this hex coordinate is one unit higher on the I-axis.
     * @see #front(Hex)
     */
    public boolean frontI(Hex other){
        return this.x == other.x + 2 && this.y == other.y - 1;
    }
    /**
     * Determines if this hex coordinate is in front of another hex coordinate on the J-axis.
     * A hex coordinate is considered "in the front" if it is positioned one step higher.
     *
     * @param other The other hex to compare with.
     * @return True if this hex coordinate is one unit higher on the J-axis.
     * @see #front(Hex)
     */
    public boolean frontJ(Hex other){
        return this.x == other.x + 1 && this.y == other.y + 1;
    }
    /**
     * Determines if this hex coordinate is in front of another hex coordinate on the K-axis.
     * A hex coordinate is considered "in the front" if it is positioned one step higher.
     *
     * @param other The other hex to compare with.
     * @return True if this hex coordinate is one unit higher on the K-axis.
     * @see #front(Hex)
     */
    public boolean frontK(Hex other){
        return this.x == other.x - 1 && this.y == other.y + 2;
    }
    /**
     * Determines if this hex coordinate is behind another hex coordinate on the I-axis
     * A hex coordinate is considered "behind" if it is positioned one step lower.
     *
     * @param other The other hex to compare with.
     * @return True if this hex coordinate is one unit lower on the I-axis.
     * @see #back(Hex)
     */
    public boolean backI(Hex other){
        return this.x == other.x - 2 && this.y == other.y + 1;
    }
    /**
     * Determines if this hex coordinate is behind another hex coordinate on the J-axis
     * A hex coordinate is considered "behind" if it is positioned one step lower.
     *
     * @param other The other hex to compare with.
     * @return True if this hex coordinate is one unit lower on the J-axis.
     * @see #back(Hex)
     */
    public boolean backJ(Hex other){
        return this.x == other.x - 1 && this.y == other.y - 1;
    }
    /**
     * Determines if this hex coordinate is behind another hex coordinate on the K-axis
     * A hex coordinate is considered "behind" if it is positioned one step lower.
     *
     * @param other The other hex to compare with.
     * @return True if this hex coordinate is one unit lower on the K-axis.
     * @see #back(Hex)
     */
    public boolean backK(Hex other){
        return this.x == other.x + 1 && this.y == other.y - 2;
    }
    /**
     * Checks if this hex is equal to another hex.
     * Two hex coordinates are considered equal if they have the same coordinates in all axes.
     *
     * @param other The other hex to compare.
     * @return True if both hex coordinates have the same coordinates, otherwise false.
     */
    public boolean equals(Hex other) {
        return this.x == other.x && this.y == other.y;
    }
    /**
     * Checks if this hex coordinate is within a given radius from the origin, as specified in the grid.
     * The radius is determined using the hexagonal distance metric.
     *
     * @param radius The radius to check.
     * @return True if the hex is within the radius, otherwise false.
     */
    public boolean inRange(int radius){
        return 0 <= getLineI() && getLineI() < radius*2 - 1 &&
                -radius < getLineJ() && getLineJ() < radius &&
                0 <= getLineK() && getLineK() < radius*2 - 1;
    }

    // convert to rectangular
    /**
     * Converts the hexagonal coordinates to a rectangular X coordinate.
     * This transformation is based on the hexagonal grid layout, where the X-coordinate
     * is computed using the sine of 60 degrees to account for the hexagonal tiling pattern.
     *
     * @return The X-coordinate in rectangular space.
     */
    public double X(){
        return sinOf60 * (x+y);
    }
    /**
     * Converts the hexagonal coordinates to a rectangular Y coordinate.
     * This transformation is based on the hexagonal grid layout, where the Y-coordinate
     * is computed using the sine of 30 degrees to account for the hexagonal tiling pattern.
     *
     * @return The Y-coordinate in rectangular space.
     */
    public double Y(){
        return (x-y)/4.0;
    }
    /**
     * String representation of the hex coordinate used for debugging
     * <p>Format: {@code {Hex I,J,K = {i, j, k}, Line I,J,K = {i, j, k}, Rect X,Y = {x, y}}}</p>
     * @return A string representation of the hex coordinate, including its rectangular conversion.
     */
    public String toString(){
        return "{Hex I,J,K = {" + I() + ", " + J() + ", " + K() +
                "}, Line I,J,K = {" + getLineI() + ", " + getLineJ() + ", " + getLineK() +
                "}, Rect X,Y = {" + X() + ", "+ Y() + "}}";
    }

    // Coordinate manipulation
    /**
     * Moves the hex coordinate along the I-axis.
     *
     * @param unit The number of units to move.
     */
    public void moveI(int unit){
        this.x += 2 * unit;
        this.y -= unit;
    }
    /**
     * Moves the hex coordinate along the J-axis.
     *
     * @param unit The number of units to move.
     */
    public void moveJ(int unit){
        this.x += unit;
        this.y += unit;
    }
    /**
     * Moves the hex coordinate along the K-axis.
     *
     * @param unit The number of units to move.
     */
    public void moveK(int unit){
        this.x -= unit;
        this.y += 2 * unit;
    }
    /**
     * Creates a new hex coordinate shifted along the I-axis.
     *
     * @param unit The number of units to shift.
     * @return A new hex coordinate shifted along the I-axis.
     */
    public Hex shiftI(int unit){
        return new Hex (this.x + 2 * unit, this.y - unit);
    }
    /**
     * Creates a new hex coordinate shifted along the J-axis.
     *
     * @param unit The number of units to shift.
     * @return A new hex coordinate shifted along the J-axis.
     */
    public Hex shiftJ(int unit){
        return new Hex (this.x + unit, this.y + unit);
    }
    /**
     * Creates a new hex coordinate shifted along the K-axis.
     *
     * @param unit The number of units to shift.
     * @return A new hex coordinate shifted along the K-axis.
     */
    public Hex shiftK(int unit){
        return new Hex (this.x - unit, this.y + 2 * unit);
    }

    // Add and subtract
    /**
     * Adds another hex to this hex coordinate and returns a new hex coordinate.
     *
     * @param other The hex coordinate to add.
     * @return A new hex coordinate with the summed coordinates.
     */
    public Hex add(Hex other){
        return new Hex(this.x + other.x, this.y + other.y);
    }
    /**
     * Subtracts another hex coordinate from this hex coordinate and returns a new hex coordinate.
     *
     * @param other The hex coordinate to subtract.
     * @return A new hex coordinate with the subtracted coordinates.
     */
    public Hex subtract(Hex other){
        return new Hex(this.x - other.x, this.y - other.y);
    }
    // Set
    /**
     * Adds set hex to another hex coordinate.
     *
     * @param other The target hex coordinate.
     */
    public void set(Hex other){
        this.x = other.x;
        this.y = other.y;
    }
    // Get
    /**
     * This hex coordinate
     *
     * @return this hex coordinate object
     */
    public Hex thisHex(){
        return Hex.hex(this.getLineI(), this.getLineK());
    }
}

File: python/hex/Hex.py
import math

class Hex:
    """
    The Hex class represents a 2D coordinate in a hexagonal grid system using
    a specialized integer coordinate model. It supports both raw coordinate access
    and derived line-based computations across three axes: I, J, and K.

    Coordinate System:
    - Designed by William Wu.
    - Axes I, J, K run diagonally through the hexagonal grid.
    - I+ is 60 degrees to J+, J+ is 60 degrees to K+, K+ is 60 degrees to J-.
    - Coordinates (i, k) correspond to a basis for representing any hexagon.
    - Raw coordinates: distance along one axis multiplied by 2, with i - j + k = 0.
    - Line coordinates: distance perpendicular to axes, with I + J - K = 0.
    - Line coordinates are preferred due to complexities with raw coordinates.

    Coordinate System Implementation:
    - x and y are base values stored in each Hex instance.
    - I = x, K = y, J = x + y.
    - Line indices: get_line_i = (2y + x) / 3, get_line_j = (x - y) / 3, get_line_k = (2x + y) / 3.

    Functionality:
    - Access raw coordinates: I(), J(), K().
    - Access line coordinates: get_line_i(), get_line_j(), get_line_k().
    - Create Hex objects: constructors or factory methods hex().
    - Move along I, J, K axes: move_i(), move_j(), move_k().
    - Addition/subtraction: add(), subtract().
    - Check line alignment and adjacency: in_line_i(), adjacent(), etc.
    - Determine relative orientation: front(), back(), and axis-specific versions.
    - Clone instances.

    Usage Notes:
    - Use factory method hex(i, k) instead of direct constructors for correct line coordinate shifts.
    """
    _half_sin_of_60 = math.sqrt(3) / 4

    def __init__(self, i=0, k=0):
        """Initialize a Hex coordinate at (i, k). Defaults to (0, 0)."""
        self._x = i
        self._y = k

    @staticmethod
    def hex(i=0, k=0):
        """Create a Hex coordinate using line indices, shifted accordingly."""
        return Hex().shift_i(k).shift_k(i)

    # Raw coordinates
    def I(self):
        """Return raw I-coordinate."""
        return self._x

    def J(self):
        """Return raw J-coordinate."""
        return self._x + self._y

    def K(self):
        """Return raw K-coordinate."""
        return self._y

    # Line coordinates
    def get_line_i(self):
        """Compute line index along I-axis."""
        return (2 * self._y + self._x) // 3

    def get_line_j(self):
        """Compute line index along J-axis."""
        return (self._x - self._y) // 3

    def get_line_k(self):
        """Compute line index along K-axis."""
        return (2 * self._x + self._y) // 3

    def get_lines(self):
        """Return string of line indices: {I = i, J = j, K = k}."""
        return f"{{I = {self.get_line_i()}, J = {self.get_line_j()}, K = {self.get_line_k()}}}"

    # Line booleans
    def in_line_i(self, line_or_hex):
        """Check if Hex is in the given I-line or same I-line as another Hex."""
        if isinstance(line_or_hex, int):
            return self.get_line_i() == line_or_hex
        return self.get_line_i() == line_or_hex.get_line_i()

    def in_line_j(self, line_or_hex):
        """Check if Hex is in the given J-line or same J-line as another Hex."""
        if isinstance(line_or_hex, int):
            return self.get_line_j() == line_or_hex
        return self.get_line_j() == line_or_hex.get_line_j()

    def in_line_k(self, line_or_hex):
        """Check if Hex is in the given K-line or same K-line as another Hex."""
        if isinstance(line_or_hex, int):
            return self.get_line_k() == line_or_hex
        return self.get_line_k() == line_or_hex.get_line_k()

    def adjacent(self, other):
        """Check if this Hex is adjacent to another Hex."""
        return self.front(other) or self.back(other)

    def front(self, other):
        """Check if this Hex is in front of another Hex on any axis."""
        return self.front_i(other) or self.front_j(other) or self.front_k(other)

    def back(self, other):
        """Check if this Hex is behind another Hex on any axis."""
        return self.back_i(other) or self.back_j(other) or self.back_k(other)

    def front_i(self, other):
        """Check if this Hex is one unit higher on I-axis."""
        return self._x == other._x + 2 and self._y == other._y - 1

    def front_j(self, other):
        """Check if this Hex is one unit higher on J-axis."""
        return self._x == other._x + 1 and self._y == other._y + 1

    def front_k(self, other):
        """Check if this Hex is one unit higher on K-axis."""
        return self._x == other._x - 1 and self._y == other._y + 2

    def back_i(self, other):
        """Check if this Hex is one unit lower on I-axis."""
        return self._x == other._x - 2 and self._y == other._y + 1

    def back_j(self, other):
        """Check if this Hex is one unit lower on J-axis."""
        return self._x == other._x - 1 and self._y == other._y - 1

    def back_k(self, other):
        """Check if this Hex is one unit lower on K-axis."""
        return self._x == other._x + 1 and self._y == other._y - 2

    def equals(self, other):
        """Check if this Hex equals another Hex."""
        return self._x == other._x and self._y == other._y

    def in_range(self, radius):
        """Check if Hex is within given radius from origin."""
        return (0 <= self.get_line_i() < radius * 2 - 1 and
                -radius < self.get_line_j() < radius and
                0 <= self.get_line_k() < radius * 2 - 1)

    # Rectangular coordinates
    def X(self):
        """Convert to rectangular X coordinate."""
        return self._half_sin_of_60 * (self._x + self._y)

    def Y(self):
        """Convert to rectangular Y coordinate."""
        return (self._x - self._y) / 4.0

    def __str__(self):
        """Return string representation for debugging."""
        return (f"Hex[raw = {{{self.I()}, {self.J()}, {self.K()}}}, "
                f"line = {{{self.get_line_i()}, {self.get_line_j()}, {self.get_line_k()}}}, "
                f"rect = {{{self.X()}, {self.Y()}}}]")

    # Coordinate manipulation
    def move_i(self, unit):
        """Move Hex along I-axis by unit."""
        self._x += 2 * unit
        self._y -= unit

    def move_j(self, unit):
        """Move Hex along J-axis by unit."""
        self._x += unit
        self._y += unit

    def move_k(self, unit):
        """Move Hex along K-axis by unit."""
        self._x -= unit
        self._y += 2 * unit

    def shift_i(self, unit):
        """Return new Hex shifted along I-axis."""
        return Hex(self._x + 2 * unit, self._y - unit)

    def shift_j(self, unit):
        """Return new Hex shifted along J-axis."""
        return Hex(self._x + unit, self._y + unit)

    def shift_k(self, unit):
        """Return new Hex shifted along K-axis."""
        return Hex(self._x - unit, self._y + 2 * unit)

    # Add and subtract
    def add(self, other):
        """Return new Hex with summed coordinates."""
        return Hex(self._x + other._x, self._y + other._y)

    def subtract(self, other):
        """Return new Hex with subtracted coordinates."""
        return Hex(self._x - other._x, self._y - other._y)

    def set(self, other):
        """Set this Hex to another's coordinates."""
        self._x = other._x
        self._y = other._y

    def this_hex(self):
        """Return this Hex as a new instance using line coordinates."""
        return Hex.hex(self.get_line_i(), self.get_line_k())

    def __copy__(self):
        """Return a copy of this Hex."""
        return Hex(self._x, self._y)

    def __deepcopy__(self, memo):
        """Return a deep copy of this Hex."""
        return Hex(self._x, self._y)
